enhancements/olm-feature-gate.md

```diff
@@ -0,0 +1,395 @@
+---
+title: feature-gate
+authors:
+  - "@fgiloux"
+reviewers:
+  - "@awgreene"
+  - "@joelanford"
+  - "@perdasilva"
+approvers:
+  - "@perdasilva"
+creation-date: 2022-04-30
+last-updated: 2022-04-30
+status: implementable
+see-also:
+  - ""
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# Feature Gate
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [olm-docs](https://github.com/operator-framework/olm-docs/)
+
+## Summary
+
+Feature gates are a set of key/value pairs designed for enabling/disabling a specific OLM feature.
+This enhancement proposal suggests to follow the [feature stages defined in Kubernetes](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages) and to make feature gates configurable in [OLMConfig](https://olm.operatorframework.io/docs/advanced-tasks/configuring-olm/) through a generic API.
+
+## Motivation
+
+[Feature Toggles](https://martinfowler.com/articles/feature-toggles.html) are a powerful technique allowing the modification of the behaviour of a system by activating new capabilities without modification of the code. Kubernetes feature gates are an implementation of this pattern.
+
+This enhancement proposal aims at bringing the following benefits to OLM:
+
+- A more phased approach of releasing features.
```

Comment:
Nothing in this EP really addresses the implications of a user turning on an alpha/beta/techpreview feature.  This is why OCP has the concept of "Techpreviewnoupgrade" among other things, in order to:

1) make it abundantly clear to the admin that the feature they are turning on is techpreview (you can also achieve this by naming your featuregates appropriately)
2) make it clear they will not be able to upgrade the cluster (And then enforcing that constraint).  It is *somewhat* up to OLM if they want a similar enforcement, but there are many many subtleties to deal with if you allow upgrades after a techpreview feature has defined some APIs and stored resources using that api.  I think it is preferable if you can tie the enablement of TP features to blocking cluster upgrades (of course that is an OCP-specific aspect, so it may not belong in upstream OLM)

One option might be for downstream OLM to override/disable any featuregates that are techpreview, unless the OCP techpreviewnoupgrade featuregate is enabled at the cluster level.

In any case, i'd like to see more discussion of how alpha/beta/techpreview featuregates will be managed/expressed.

Reply:
(I assume you're familiar, but in case you're not: https://docs.openshift.com/container-platform/4.10/nodes/clusters/nodes-cluster-enabling-features.html)